### PR TITLE
Remove dependency on expressjs request object from API

### PR DESCRIPTION
**Pull request by clark800**

### CODE REVIEW

File: api/lib/validate.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+const _ = require('lodash');
+const errors = require('./errors.js');
+const validator = require('./schema-validator');
+const ripple = require('ripple-lib');
+
+function addOptionalOption(validator) {
```
Comment (lumberj) [inline api/lib/validate.js]: Aren't all Options optional?

File: api/lib/validate.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    if (optional && parameter === undefined) {
+      return null;
+    }
+    return validator(parameter);
+  };
+}
+
+function validateAccount(account) {
+  if (!ripple.UInt160.is_valid(account)) {
+    return new errors.InvalidRequestError(
```
Comment (lumberj) [inline api/lib/validate.js]: Is this meant to be part of `/api`? The error messages here are for the server

File: api/lib/validate.js (hunk 1 above)
Comment (clark800) [inline api/lib/validate.js]: This is a tricky one to name. It is adding an option that represents whether the value is optional.

File: api/lib/validate.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+function validateCounterparty(counterparty) {
+  if (!ripple.UInt160.is_valid(counterparty)) {
+    return new errors.InvalidRequestError(
+      'Parameter is not a valid Ripple address: counterparty');
+  }
+  return null;
+}
+
+function validateIssue(issue) {
```
Comment (lumberj) [inline api/lib/validate.js]: Good naming

Comment (clark800) [inline api/lib/validate.js]: Got that from Vinnie, he said they spent a week coming up with it!

File: api/balances.js [hunk 4]
```diff
@@ -28,67 +29,49 @@ const DefaultPageLimit = 200;
  *  @param {Number String} [request.query.ledger] - identifier
```
Comment (lumberj) [inline api/balances.js]: Can these JSDoc comments stay updated?

File: api/lib/validate.js (hunk 2 above)
Comment (clark800) [inline api/lib/validate.js]: Yes, it is part of `api`. We should rename the error classes, but I think that can be a separate pull request.

File: api/balances.js (hunk 4 above)
Comment (clark800) [inline api/balances.js]: Yes, but I didn't want to update them until everyone agrees on the calling syntax.

File: api/lib/tx-to-rest-converter.js [hunk 5]
```diff
@@ -351,24 +351,22 @@ const AccountSetResponseFlags = {
...
 
-TxToRestConverter.prototype.parseSettingResponseFromTx = function(params, message, meta, callback) {
+TxToRestConverter.prototype.parseSettingResponseFromTx = function(settings, message, meta, callback) {
   var result = {
     settings: {}
   };
 
-  // lazy loading to avoid circular dependency
-  var settings = require('../settings.js');
+  // cannot require at the top due to circular dependency
```
Comment (lumberj) [inline api/lib/tx-to-rest-converter.js]: Isn't it being require'd at the top though?

Comment (clark800) [inline api/lib/tx-to-rest-converter.js]: Yeah, but it fails if you take this out.

**Comment by lumberj** [comment]
So far, preliminary-LGTM. No changes in the tests makes me confident in these changes

File: api/lib/validate.js (hunk 2 above)
Comment (geertweening) [inline api/lib/validate.js]: +1 they are actual logic errors, the validation fails, which is an api thing

File: api/lib/validate.js (hunk 1 above)
Comment (geertweening) [inline api/lib/validate.js]: Very confusing, add comments to make it more clear?

File: server/routes.js [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+function getOrderBook(request, callback) {
+  const account = request.params.account;
+  const base = request.params.base;
+  const counter = request.params.counter;
+  const options = {
+    ledger: request.query.ledger,
+    limit: request.query.limit,
+    marker: request.query.marker
+  };
+  api.orders.getOrderBook(account, base, counter, options, callback);
```
Comment (geertweening) [inline server/routes.js]: I like an `(options, callback)` scheme better where all these arguments are captured in the options parameter except for the callback

Comment (lumberj) [inline server/routes.js]: Those parameters are never optional

Comment (lumberj) [inline server/routes.js]: From Clean Code:

> The ideal number of arguments for a function is zero (niladic). Next comes one (monadic), followed closely by two (dyadic). Three arguments (triadic) should be avoided where possible. More than three (polyadic) requires very special justification -- and then shouldn't be used anyway.

Also: http://refactoring.com/catalog/introduceParameterObject.html

Comment (lumberj) [inline server/routes.js]: Since this PR is not meant to be a full refactoring, we should address this in a later PR. We're already circulating a proposal for a refactor of the high level API, and we can comment there.

**Comment by lumberj** [comment]
LGTM

File: server/routes.js (hunk 6 above)
Comment (geertweening) [inline server/routes.js]: Makes sense to me

**Comment by geertweening** [comment]
LGTM
